File: server/evaluation/wrappers/operation-evaluator.ts
/**
 * A7: Operation Evaluator - TypeScript Wrapper
 *
 * Phase 5: Refactored to use real AppsFlyer cohort data.
 * The Python-based evaluation is kept for backward compatibility but deprecated.
 */

import { spawn } from 'child_process'
import path from 'path'
import { and, eq, gte, lte, sql } from 'drizzle-orm'
import { db } from '../../db'
import {
  campaigns,
  changeEvents,
  type ChangeEvent,
  type NewOperationScore,
} from '../../db/schema'
import {
  getOperationCohortMetrics,
  createOperationScore,
} from '../../db/queries-evaluation'
import { calculateBaselineRoas, calculateBaselineRetention } from '../../db/queries-appsflyer'
import { getOrCreateBaselineSettings } from './baseline-calculator'

// ============================================
// TYPE DEFINITIONS
// ============================================

export type ScoreStage = 'T+1' | 'T+3' | 'T+7'

export interface OperationStageResult {
  stage: ScoreStage
  stageDays: number
  baseScore: number | null
  finalScore: number | null
  minAchievement: number | null
  roasAchievement: number | null
  retentionAchievement: number | null
  riskLevel: 'danger' | 'warning' | 'observe' | 'healthy' | 'excellent' | null
  operationScoreId?: number
  dataStatus: 'complete' | 'pending' | 'missing'
  error?: string
}

export interface OperationEvaluationResult {
  operationId: number
  campaignId: string
  campaignName?: string
  optimizerEmail: string
  operationType: string
  operationDate: string
  stages: OperationStageResult[]
  dataSource?: 'appsflyer' | 'mock'
  context?: {
    appId: string
    geo: string
    mediaSource: string
    campaignKey: string
  }
}

export interface OptimizerStats {
  optimizer_email: string
  total_operations: number
  avg_roas_achievement: number
  avg_ret_achievement: number
  avg_min_achievement: number
  excellent_count: number
  excellent_rate: number
  good_count: number
  good_rate: number
  failed_count: number
  failed_rate: number
}

export interface LeaderboardResult {
  period_days: number
  start_date: string
  end_date: string
  total_optimizers: number
  leaderboard: OptimizerStats[]
  error?: string
  dataSource?: 'appsflyer' | 'mock'
}

export interface BatchEvaluationResult {
  success: boolean
  target_date: string
  total_operations: number
  success_count: number
  failed_count: number
  pending_count: number
  results: Array<{
    operation_id: number
    optimizer?: string
    stages?: OperationStageResult[]
    error?: string
    pending?: boolean
  }>
  error?: string
  dataSource?: 'appsflyer' | 'mock'
}

// ============================================
// HELPER FUNCTIONS
// ============================================

const STAGE_DAY_MAP: Record<ScoreStage, number> = {
  'T+1': 1,
  'T+3': 3,
  'T+7': 7,
}

const STAGE_FACTOR: Record<ScoreStage, number> = {
  'T+1': 0.5,
  'T+3': 0.8,
  'T+7': 1,
}

const DEFAULT_STAGES: ScoreStage[] = ['T+1', 'T+3', 'T+7']

function mapToBaseScore(achievement: number | null): number | null {
  if (achievement === null || Number.isNaN(achievement)) return null
  if (achievement < 0.6) return 0
  if (achievement < 0.85) return 40
  if (achievement < 1.0) return 60
  if (achievement < 1.1) return 80
  return 100
}

function getRiskLevel(
  achievement: number | null
): OperationStageResult['riskLevel'] {
  if (achievement === null || Number.isNaN(achievement)) return null
  if (achievement < 0.6) return 'danger'
  if (achievement < 0.85) return 'warning'
  if (achievement < 1.0) return 'observe'
  if (achievement < 1.1) return 'healthy'
  return 'excellent'
}

function mapSuggestionType(riskLevel: OperationStageResult['riskLevel']): string | null {
  switch (riskLevel) {
    case 'danger':
      return 'stop'
    case 'warning':
      return 'shrink'
    case 'observe':
      return 'observe'
    case 'healthy':
    case 'excellent':
      return 'expand'
    default:
      return null
  }
}

function classifyOperationMagnitude(changePercentage: number | null | undefined) {
  if (changePercentage === null || changePercentage === undefined) return { magnitude: null, label: null }
  const absChange = Math.abs(changePercentage)
  if (absChange <= 0.05) return { magnitude: absChange, label: '微调' }
  if (absChange <= 0.2) return { magnitude: absChange, label: '常规调整' }
  return { magnitude: absChange, label: '大胆操作' }
}

// ============================================
// APPSFLYER-BASED EVALUATION (PRIMARY)
// ============================================
interface OperationContext {
  changeEvent: ChangeEvent
  campaignId: string
  campaignName?: string
  appId: string
  geo: string
  mediaSource: string
  campaignKey: string
}

async function pickAfContext(campaignIdentifiers: string[]): Promise<{
  appId: string
  geo: string
  mediaSource: string
  campaignKey: string
} | null> {
  for (const identifier of campaignIdentifiers) {
    const res = await db.execute(sql`
      SELECT app_id, geo, media_source, SUM(installs) AS installs
      FROM af_cohort_kpi_daily
      WHERE campaign = ${identifier} AND days_since_install = 0
      GROUP BY app_id, geo, media_source
      ORDER BY SUM(installs) DESC
      LIMIT 1
    `)

    const row = res.rows[0] as Record<string, unknown> | undefined
    if (row) {
      return {
        appId: String(row.app_id),
        geo: String(row.geo),
        mediaSource: String(row.media_source),
        campaignKey: identifier,
      }
    }
  }
  return null
}

async function resolveOperationContext(operationId: number): Promise<OperationContext> {
  const [operation] = await db
    .select()
    .from(changeEvents)
    .where(eq(changeEvents.id, operationId))
    .limit(1)

  if (!operation) {
    throw new Error(`Operation ${operationId} not found`)
  }

  const campaignId = operation.campaign || operation.resourceName

  let campaignName: string | undefined
  if (campaignId) {
    const campaignRow = await db
      .select({ name: campaigns.name })
      .from(campaigns)
      .where(and(eq(campaigns.accountId, operation.accountId), eq(campaigns.resourceName, campaignId)))
      .limit(1)

    campaignName = campaignRow[0]?.name || undefined
  }

  const campaignIdentifiers = [campaignId, campaignName].filter(Boolean) as string[]
  const afContext = await pickAfContext(campaignIdentifiers)

  const fallbackAppId = process.env.AF_APP_ID
  const fallbackGeo = process.env.AF_DEFAULT_GEO
  const fallbackMediaSource = process.env.AF_DEFAULT_MEDIA_SOURCE

  const appId = afContext?.appId || fallbackAppId
  const geo = afContext?.geo || fallbackGeo
  const mediaSource = afContext?.mediaSource || fallbackMediaSource
  const campaignKey = afContext?.campaignKey || campaignIdentifiers[0]

  if (!appId || !geo || !mediaSource || !campaignKey) {
    throw new Error('Missing AppsFlyer context for operation scoring')
  }

  return {
    changeEvent: operation,
    campaignId,
    campaignName,
    appId,
    geo,
    mediaSource,
    campaignKey,
  }
}

async function evaluateStage(params: {
  context: OperationContext
  stage: ScoreStage
  baselineDays: number
}): Promise<OperationStageResult & { operationScoreId?: number }> {
  const { context, stage, baselineDays } = params
  const stageDays = STAGE_DAY_MAP[stage]

  const operationDate = new Date(context.changeEvent.timestamp)

  const metrics = await getOperationCohortMetrics({
    campaign: context.campaignKey,
    operationDate,
    appId: context.appId,
    geo: context.geo,
    mediaSource: context.mediaSource,
    daysSinceInstall: stageDays,
  })

  if (!metrics) {
    return {
      stage,
      stageDays,
      baseScore: null,
      finalScore: null,
      minAchievement: null,
      roasAchievement: null,
      retentionAchievement: null,
      riskLevel: null,
      dataStatus: 'pending',
      error: 'Insufficient cohort data for this stage',
    }
  }

  const [baselineRoas, baselineRet] = await Promise.all([
    calculateBaselineRoas({
      appId: context.appId,
      geo: context.geo,
      mediaSource: context.mediaSource,
      baselineDays,
      daysSinceInstall: stageDays,
    }),
    calculateBaselineRetention({
      appId: context.appId,
      geo: context.geo,
      mediaSource: context.mediaSource,
      daysSinceInstall: stageDays,
      baselineDays,
    }),
  ])

  const roasAchievement = baselineRoas && baselineRoas > 0 && metrics.actualRoas7 !== null
    ? metrics.actualRoas7 / baselineRoas
    : null

  const retentionAchievement = baselineRet && baselineRet > 0 && metrics.actualRet7 !== null
    ? metrics.actualRet7 / baselineRet
    : null

  const minAchievement = [roasAchievement, retentionAchievement]
    .filter((v): v is number => v !== null && !Number.isNaN(v))
    .reduce<number | null>((acc, curr) => {
      if (acc === null) return curr
      return Math.min(acc, curr)
    }, null)

  const baseScore = mapToBaseScore(minAchievement)
  const stageFactor = STAGE_FACTOR[stage]
  const finalScore = baseScore !== null ? baseScore * stageFactor : null
  const riskLevel = getRiskLevel(minAchievement)
  const suggestionType = mapSuggestionType(riskLevel)

  // Operation magnitude (optional)
  const changePct = (
    context.changeEvent.fieldChanges as Record<string, unknown> | null | undefined
  )?.change_percentage as number | undefined
  const operationMagnitude = classifyOperationMagnitude(changePct ?? null)

  // Score date = operation date + stageDays
  const evaluationDate = new Date(operationDate)
  evaluationDate.setDate(evaluationDate.getDate() + stageDays)

  let savedScoreId: number | undefined
  const shouldPersist = metrics !== null && baseScore !== null

  if (shouldPersist) {
    const payload: NewOperationScore = {
      operationId: context.changeEvent.id,
      campaignId: context.campaignId,
      optimizerEmail: context.changeEvent.userEmail,
      operationType: context.changeEvent.operationType,
      operationDate: operationDate.toISOString().split('T')[0],
      evaluationDate: evaluationDate.toISOString().split('T')[0],
      scoreStage: stage,
      stageFactor: stageFactor.toString(),
      actualRoas: metrics.actualRoas7 !== null ? metrics.actualRoas7.toString() : null,
      actualRet: metrics.actualRet7 !== null ? metrics.actualRet7.toString() : null,
      baselineRoas: baselineRoas !== null && baselineRoas !== undefined ? baselineRoas.toString() : null,
      baselineRet: baselineRet !== null && baselineRet !== undefined ? baselineRet.toString() : null,
      roasAchievement: roasAchievement !== null ? roasAchievement.toString() : null,
      retentionAchievement: retentionAchievement !== null ? retentionAchievement.toString() : null,
      minAchievement: minAchievement !== null ? minAchievement.toString() : null,
      roasAchievementRate: roasAchievement !== null ? (roasAchievement * 100).toString() : null,
      retAchievementRate: retentionAchievement !== null ? (retentionAchievement * 100).toString() : null,
      riskLevel,
      baseScore,
      finalScore: finalScore !== null ? Number(finalScore).toFixed(2) : null,
      valueBefore: null,
      valueAfter: null,
      changePercentage: changePct !== undefined && changePct !== null ? changePct.toString() : null,
      operationMagnitude: operationMagnitude.magnitude !== null ? operationMagnitude.magnitude.toString() : null,
      operationTypeLabel: operationMagnitude.label,
      isBoldSuccess: baseScore !== null && baseScore >= 80 && (operationMagnitude.magnitude || 0) > 0.2,
      specialRecognition:
        baseScore !== null && baseScore >= 80 && (operationMagnitude.magnitude || 0) > 0.2
          ? '🌟 大胆创新奖'
          : baseScore !== null && baseScore >= 80 && (operationMagnitude.magnitude || 0) <= 0.05
            ? '🎯 精准调优奖'
            : baseScore !== null && baseScore >= 100
              ? '🏆 卓越表现奖'
              : null,
      suggestionType,
      suggestionDetail: suggestionType ? `Auto-suggest: ${suggestionType}` : null,

      // Legacy columns for compatibility
      actualRoas7: metrics.actualRoas7 !== null ? metrics.actualRoas7.toString() : null,
      actualRet7: metrics.actualRet7 !== null ? metrics.actualRet7.toString() : null,
      baselineRoas7: baselineRoas !== null && baselineRoas !== undefined ? baselineRoas.toString() : null,
      baselineRet7: baselineRet !== null && baselineRet !== undefined ? baselineRet.toString() : null,
    }

    const saved = await createOperationScore(payload)
    savedScoreId = saved?.id
  }

  return {
    stage,
    stageDays,
    baseScore,
    finalScore,
    minAchievement,
    roasAchievement,
    retentionAchievement,
    riskLevel,
    operationScoreId: savedScoreId,
    dataStatus: baseScore === null ? 'missing' : 'complete',
    error:
      baseScore === null
        ? 'Baseline missing or insufficient to compute score'
        : undefined,
  }
}

export async function evaluateOperationFromAF(params: {
  operationId: number
  stages?: ScoreStage[]
}): Promise<OperationEvaluationResult> {
  const { operationId, stages = DEFAULT_STAGES } = params

  try {
    const context = await resolveOperationContext(operationId)
    const baselineSettings = await getOrCreateBaselineSettings({
      appId: context.appId,
      geo: context.geo,
      mediaSource: context.mediaSource,
    })

    const stageResults: OperationStageResult[] = []

    for (const stage of stages) {
      const result = await evaluateStage({
        context,
        stage,
        baselineDays: baselineSettings.baselineDays,
      })
      stageResults.push(result)
    }

    // Update change_events.operation_scores for quick lookup
    await db
      .update(changeEvents)
      .set({
        operationScores: stageResults.map((stage) => ({
          stage: stage.stage,
          finalScore: stage.finalScore,
          baseScore: stage.baseScore,
          minAchievement: stage.minAchievement,
          riskLevel: stage.riskLevel,
          operationScoreId: stage.operationScoreId,
          dataStatus: stage.dataStatus,
          error: stage.error,
        })),
      })
      .where(eq(changeEvents.id, operationId))

    return {
      operationId,
      campaignId: context.campaignId,
      campaignName: context.campaignName,
      optimizerEmail: context.changeEvent.userEmail,
      operationType: context.changeEvent.operationType,
      operationDate: new Date(context.changeEvent.timestamp).toISOString().split('T')[0],
      stages: stageResults,
      dataSource: 'appsflyer',
      context: {
        appId: context.appId,
        geo: context.geo,
        mediaSource: context.mediaSource,
        campaignKey: context.campaignKey,
      },
    }
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown error'
    return {
      operationId,
      campaignId: '',
      optimizerEmail: '',
      operationType: '',
      operationDate: '',
      stages: [
        {
          stage: 'T+7',
          stageDays: 7,
          baseScore: null,
          finalScore: null,
          minAchievement: null,
          roasAchievement: null,
          retentionAchievement: null,
          riskLevel: null,
          dataStatus: 'missing',
          error: message,
        },
      ],
      dataSource: 'appsflyer',
    }
  }
}

/**
 * Batch evaluate all operations from 7 days ago using AppsFlyer data
 *
 * This should be run daily to evaluate operations that were made 7 days ago.
 * Requires change_events table to track operations.
 *
 * NOTE: Current change_events schema doesn't have appId/geo/mediaSource fields.
 * This function is designed for future use when change_events is extended with
 * AppsFlyer-related fields. Currently returns empty results with a note.
 *
 * @param targetDate - The date to evaluate operations for (default: 7 days ago)
 * @returns Batch evaluation summary
 */
export async function evaluateOperations7DaysAgoFromAF(
  targetDate?: Date
): Promise<BatchEvaluationResult> {
  // Target date is 7 days ago by default
  const target = targetDate || new Date()
  if (!targetDate) {
    target.setDate(target.getDate() - 7)
  }

  // Get operations from the target date using timestamp field
  const startOfDay = new Date(target)
  startOfDay.setHours(0, 0, 0, 0)
  const endOfDay = new Date(target)
  endOfDay.setHours(23, 59, 59, 999)

  try {
    const operations = await db
      .select()
      .from(changeEvents)
      .where(
        and(
          gte(changeEvents.timestamp, startOfDay),
          lte(changeEvents.timestamp, endOfDay),
          eq(changeEvents.resourceType, 'campaign')
        )
      )

    const results: BatchEvaluationResult['results'] = []
    let successCount = 0
    let failedCount = 0
    let pendingCount = 0

    for (const op of operations) {
      const evaluation = await evaluateOperationFromAF({ operationId: op.id, stages: ['T+7'] })
      const stage = evaluation.stages[0]

      if (stage.dataStatus === 'complete') {
        successCount++
      } else if (stage.dataStatus === 'pending') {
        pendingCount++
      } else {
        failedCount++
      }

      results.push({
        operation_id: op.id,
        optimizer: op.userEmail,
        stages: evaluation.stages,
        error: stage.error,
        pending: stage.dataStatus === 'pending',
      })
    }

    return {
      success: failedCount === 0,
      target_date: target.toISOString().split('T')[0],
      total_operations: operations.length,
      success_count: successCount,
      failed_count: failedCount,
      pending_count: pendingCount,
      results,
      dataSource: 'appsflyer',
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error'
    return {
      success: false,
      target_date: target.toISOString().split('T')[0],
      total_operations: 0,
      success_count: 0,
      failed_count: 0,
      pending_count: 0,
      results: [],
      error: `Batch evaluation failed: ${errorMessage}`,
      dataSource: 'appsflyer',
    }
  }
}

// ============================================
// LEGACY PYTHON-BASED FUNCTIONS (DEPRECATED)
// ============================================

/**
 * Run Python script and return parsed JSON output
 *
 * @deprecated Use evaluateOperationFromAF instead
 */
async function runPythonScript<T>(scriptName: string, input: Record<string, unknown>): Promise<T> {
  console.warn(
    '[DEPRECATED] Using Python-based operation evaluation. Switch to evaluateOperationFromAF for real AppsFlyer data.'
  )

  return new Promise((resolve, reject) => {
    const scriptPath = path.join(process.cwd(), 'server', 'evaluation', 'python', scriptName)

    const pythonProcess = spawn('python3', [scriptPath])

    let stdoutData = ''
    let stderrData = ''

    pythonProcess.stdout.on('data', (data) => {
      stdoutData += data.toString()
    })

    pythonProcess.stderr.on('data', (data) => {
      stderrData += data.toString()
    })

    pythonProcess.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`Python script exited with code ${code}\nStderr: ${stderrData}`))
        return
      }

      try {
        const result = JSON.parse(stdoutData)
        resolve(result)
      } catch (error) {
        reject(
          new Error(
            `Failed to parse Python output: ${error}\nOutput: ${stdoutData}\nStderr: ${stderrData}`
          )
        )
      }
    })

    pythonProcess.on('error', (error) => {
      reject(new Error(`Failed to start Python process: ${error.message}`))
    })

    pythonProcess.stdin.write(JSON.stringify(input))
    pythonProcess.stdin.end()
  })
}

/**
 * Evaluate an operation 7 days after execution
 *
 * @deprecated Use evaluateOperationFromAF instead. This function uses mock data via Python.
 *
 * Calculates ROAS7/RET7 achievement rates 7 days post-operation and scores it.
 *
 * @param operationId - Operation ID from change_events table
 * @returns Operation evaluation result with score
 */
export async function evaluateOperation(operationId: number): Promise<OperationEvaluationResult> {
  console.warn(
    '[DEPRECATED] evaluateOperation uses mock data. Use evaluateOperationFromAF instead.'
  )

  return evaluateOperationFromAF({ operationId })
}

/**
 * Get optimizer leaderboard
 *
 * @deprecated Use getOptimizerLeaderboardFromDb instead.
 *
 * Returns ranked list of optimizers based on their operation performance
 *
 * @param days - Look back period in days (default: 30)
 * @param limit - Maximum number of optimizers to return (default: 20)
 * @returns Leaderboard with optimizer statistics
 */
export async function getOptimizerLeaderboard(
  days: number = 30,
  limit: number = 20
): Promise<LeaderboardResult> {
  console.warn(
    '[DEPRECATED] getOptimizerLeaderboard uses mock data. Use getOptimizerLeaderboardFromDb instead.'
  )

  const input = {
    action: 'leaderboard',
    days,
    limit,
  }

  const result = await runPythonScript<LeaderboardResult>('operation_evaluator.py', input)
  return { ...result, dataSource: 'mock' }
}

/**
 * Batch evaluate all operations from 7 days ago
 *
 * @deprecated Use evaluateOperations7DaysAgoFromAF instead. This function uses mock data via Python.
 *
 * This should be run daily to evaluate operations that were made 7 days ago.
 *
 * @returns Batch evaluation summary
 */
export async function evaluateOperations7DaysAgo(): Promise<BatchEvaluationResult> {
  console.warn(
    '[DEPRECATED] evaluateOperations7DaysAgo uses mock data. Use evaluateOperations7DaysAgoFromAF instead.'
  )

  const input = {
    action: 'evaluate_7days_ago',
  }

  const result = await runPythonScript<BatchEvaluationResult>('operation_evaluator.py', input)
  return { ...result, dataSource: 'mock', pending_count: 0 }
}

/**
 * Re-export database functions for direct access
 */
export {
  getOperationScores,
  getOptimizerLeaderboard as getOptimizerLeaderboardFromDb,
} from '../../db/queries-evaluation'

File: server/evaluation/python/operation_evaluator.py
#!/usr/bin/env python3
"""
A5: Operation Evaluator

Evaluates optimizer operations 7 days after execution.
Generates optimizer leaderboards based on operation performance.
"""

import sys
from datetime import datetime, timedelta
from typing import Dict, Any, List, Optional
from db_utils import get_db, format_output, read_input


class OperationEvaluator:
    """Evaluate optimizer operations and generate leaderboards"""

    def __init__(self):
        self.db = get_db()

    def evaluate_operation(self, operation_id: int) -> Dict[str, Any]:
        """
        Evaluate an operation 7 days after execution

        Args:
            operation_id: Operation ID from change_events table

        Returns:
            Dictionary containing:
            - operation_id: int
            - campaign_id: str
            - optimizer_email: str
            - operation_type: str
            - operation_date: str
            - evaluation_date: str
            - actual_roas7: float
            - actual_ret7: float
            - baseline_roas7: float
            - baseline_ret7: float
            - roas_achievement_rate: float
            - ret_achievement_rate: float
            - min_achievement_rate: float
            - score: "优秀" | "合格" | "失败"
        """
        try:
            with self.db:
                # 1. Get operation record
                operation_query = """
                    SELECT
                        id,
                        user_email,
                        resource_type,
                        summary,
                        timestamp,
                        resource_name as campaign_name,
                        campaign as campaign_id
                    FROM change_events
                    WHERE id = %s
                    LIMIT 1
                """

                operation_results = self.db.execute_query(operation_query, (operation_id,))

                if not operation_results or len(operation_results) == 0:
                    return {
                        "error": f"Operation {operation_id} not found",
                        "operation_id": operation_id
                    }

                operation = operation_results[0]

                operation_date = operation['timestamp']
                evaluation_date = operation_date + timedelta(days=7)
                evaluation_date_str = evaluation_date.strftime("%Y-%m-%d")
                operation_date_str = operation_date.strftime("%Y-%m-%d")

                campaign_id = operation['campaign_id']
                optimizer_email = operation['user_email']

                # Determine operation type from summary
                summary = operation['summary']
                if 'Budget' in summary:
                    operation_type = 'BUDGET_UPDATE'
                elif 'tROAS' in summary or 'ROAS' in summary:
                    operation_type = 'TROAS_UPDATE'
                elif 'status' in summary.lower():
                    operation_type = 'STATUS_CHANGE'
                else:
                    operation_type = 'OTHER'

                # 2. Get campaign performance data 7 days after operation
                performance_query = """
                    SELECT
                        actual_roas7,
                        actual_ret7,
                        product_name,
                        country_code,
                        platform,
                        channel
                    FROM mock_campaign_performance
                    WHERE campaign_id = %s
                      AND date = %s
                    LIMIT 1
                """

                performance_results = self.db.execute_query(
                    performance_query,
                    (campaign_id, evaluation_date_str)
                )

                if not performance_results or len(performance_results) == 0:
                    return {
                        "error": f"No performance data found for campaign {campaign_id} on {evaluation_date_str}",
                        "operation_id": operation_id
                    }

                performance = performance_results[0]
                actual_roas7 = float(performance['actual_roas7'])
                actual_ret7 = float(performance['actual_ret7'])

                # 3. Get baseline at operation time
                baseline_query = """
                    SELECT baseline_roas7, baseline_ret7
                    FROM safety_baseline
                    WHERE product_name = %s
                      AND country_code = %s
                      AND platform = %s
                      AND channel = %s
                    LIMIT 1
                """

                baseline_results = self.db.execute_query(
                    baseline_query,
                    (performance['product_name'], performance['country_code'],
                     performance['platform'], performance['channel'])
                )

                if not baseline_results or len(baseline_results) == 0:
                    return {
                        "error": "No baseline found",
                        "operation_id": operation_id
                    }

                baseline = baseline_results[0]
                baseline_roas7 = float(baseline['baseline_roas7'])
                baseline_ret7 = float(baseline['baseline_ret7'])

                # 4. Calculate achievement rates
                roas_achievement_rate = (actual_roas7 / baseline_roas7 * 100) if baseline_roas7 > 0 else 0.0
                ret_achievement_rate = (actual_ret7 / baseline_ret7 * 100) if baseline_ret7 > 0 else 0.0
                min_achievement_rate = min(roas_achievement_rate, ret_achievement_rate)

                # 5. Determine score
                if min_achievement_rate >= 110:
                    score = "优秀"
                elif min_achievement_rate >= 85:
                    score = "合格"
                else:
                    score = "失败"

                # 6. Save to database
                self.save_operation_score(
                    operation_id=operation_id,
                    campaign_id=campaign_id,
                    optimizer_email=optimizer_email,
                    operation_type=operation_type,
                    operation_date=operation_date_str,
                    evaluation_date=evaluation_date_str,
                    actual_roas7=actual_roas7,
                    actual_ret7=actual_ret7,
                    baseline_roas7=baseline_roas7,
                    baseline_ret7=baseline_ret7,
                    roas_achievement_rate=roas_achievement_rate,
                    ret_achievement_rate=ret_achievement_rate
                )

                return {
                    "operation_id": operation_id,
                    "campaign_id": campaign_id,
                    "campaign_name": operation['campaign_name'],
                    "optimizer_email": optimizer_email,
                    "operation_type": operation_type,
                    "operation_date": operation_date_str,
                    "evaluation_date": evaluation_date_str,
                    "actual_roas7": actual_roas7,
                    "actual_ret7": actual_ret7,
                    "baseline_roas7": baseline_roas7,
                    "baseline_ret7": baseline_ret7,
                    "roas_achievement_rate": round(roas_achievement_rate, 2),
                    "ret_achievement_rate": round(ret_achievement_rate, 2),
                    "min_achievement_rate": round(min_achievement_rate, 2),
                    "score": score
                }

        except Exception as e:
            print(f"Operation evaluation error: {e}", file=sys.stderr, flush=True)
            return {
                "error": str(e),
                "operation_id": operation_id
            }

    def save_operation_score(
        self,
        operation_id: int,
        campaign_id: str,
        optimizer_email: str,
        operation_type: str,
        operation_date: str,
        evaluation_date: str,
        actual_roas7: float,
        actual_ret7: float,
        baseline_roas7: float,
        baseline_ret7: float,
        roas_achievement_rate: float,
        ret_achievement_rate: float
    ) -> bool:
        """Save operation score to database"""
        try:
            query = """
                INSERT INTO operation_score (
                    operation_id, campaign_id, optimizer_email, operation_type,
                    operation_date, evaluation_date,
                    actual_roas7, actual_ret7,
                    baseline_roas7, baseline_ret7,
                    roas_achievement_rate, ret_achievement_rate,
                    created_at
                )
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, NOW())
            """

            return self.db.execute_update(
                query,
                (operation_id, campaign_id, optimizer_email, operation_type,
                 operation_date, evaluation_date,
                 actual_roas7, actual_ret7,
                 baseline_roas7, baseline_ret7,
                 roas_achievement_rate, ret_achievement_rate)
            )

        except Exception as e:
            print(f"Save operation score error: {e}", file=sys.stderr, flush=True)
            return False

    def get_optimizer_leaderboard(
        self,
        days: int = 30,
        limit: int = 20
    ) -> Dict[str, Any]:
        """
        Get optimizer leaderboard based on operation performance

        Args:
            days: Look back period in days (default: 30)
            limit: Maximum number of optimizers to return (default: 20)

        Returns:
            Dictionary containing:
            - period_days: int
            - leaderboard: List of optimizer stats
        """
        try:
            with self.db:
                # Calculate date range
                end_date = datetime.now()
                start_date = end_date - timedelta(days=days)
                start_date_str = start_date.strftime("%Y-%m-%d")

                # Get optimizer statistics
                leaderboard_query = """
                    SELECT
                        optimizer_email,
                        COUNT(*) as total_operations,
                        AVG(roas_achievement_rate) as avg_roas_achievement,
                        AVG(ret_achievement_rate) as avg_ret_achievement,
                        AVG(LEAST(roas_achievement_rate, ret_achievement_rate)) as avg_min_achievement,
                        COUNT(CASE WHEN LEAST(roas_achievement_rate, ret_achievement_rate) >= 110 THEN 1 END) as excellent_count,
                        COUNT(CASE WHEN LEAST(roas_achievement_rate, ret_achievement_rate) >= 85 AND LEAST(roas_achievement_rate, ret_achievement_rate) < 110 THEN 1 END) as good_count,
                        COUNT(CASE WHEN LEAST(roas_achievement_rate, ret_achievement_rate) < 85 THEN 1 END) as failed_count
                    FROM operation_score
                    WHERE operation_date >= %s
                    GROUP BY optimizer_email
                    ORDER BY avg_min_achievement DESC
                    LIMIT %s
                """

                leaderboard = self.db.execute_query(
                    leaderboard_query,
                    (start_date_str, limit)
                )

                # Format results
                results = []
                for row in leaderboard:
                    total_ops = int(row['total_operations'])
                    excellent = int(row['excellent_count'])
                    good = int(row['good_count'])
                    failed = int(row['failed_count'])

                    results.append({
                        "optimizer_email": row['optimizer_email'],
                        "total_operations": total_ops,
                        "avg_roas_achievement": round(float(row['avg_roas_achievement']), 2),
                        "avg_ret_achievement": round(float(row['avg_ret_achievement']), 2),
                        "avg_min_achievement": round(float(row['avg_min_achievement']), 2),
                        "excellent_count": excellent,
                        "excellent_rate": round(excellent / total_ops * 100, 1) if total_ops > 0 else 0.0,
                        "good_count": good,
                        "good_rate": round(good / total_ops * 100, 1) if total_ops > 0 else 0.0,
                        "failed_count": failed,
                        "failed_rate": round(failed / total_ops * 100, 1) if total_ops > 0 else 0.0
                    })

                return {
                    "period_days": days,
                    "start_date": start_date_str,
                    "end_date": end_date.strftime("%Y-%m-%d"),
                    "total_optimizers": len(results),
                    "leaderboard": results
                }

        except Exception as e:
            print(f"Leaderboard error: {e}", file=sys.stderr, flush=True)
            return {
                "error": str(e)
            }

    def evaluate_operations_7days_ago(self) -> Dict[str, Any]:
        """
        Batch evaluate all operations from 7 days ago

        This should be run daily to evaluate operations that were made 7 days ago.

        Returns:
            Summary of evaluation results
        """
        try:
            # Calculate target date (7 days ago)
            target_date = datetime.now() - timedelta(days=7)
            target_date_str = target_date.strftime("%Y-%m-%d")

            with self.db:
                # Get all operations from 7 days ago that haven't been evaluated
                operations_query = """
                    SELECT ce.id
                    FROM change_events ce
                    LEFT JOIN operation_score os ON ce.id = os.operation_id
                    WHERE DATE(ce.timestamp) = %s
                      AND os.id IS NULL
                      AND ce.resource_type IN ('CAMPAIGN_BUDGET', 'CAMPAIGN')
                    ORDER BY ce.id
                """

                operations = self.db.execute_query(operations_query, (target_date_str,))

                results = []
                success_count = 0
                failed_count = 0

                for op in operations:
                    operation_id = op['id']
                    evaluation = self.evaluate_operation(operation_id)

                    if 'error' not in evaluation:
                        success_count += 1
                        results.append({
                            "operation_id": operation_id,
                            "optimizer": evaluation['optimizer_email'],
                            "score": evaluation['score'],
                            "min_achievement_rate": evaluation['min_achievement_rate']
                        })
                    else:
                        failed_count += 1
                        results.append({
                            "operation_id": operation_id,
                            "error": evaluation['error']
                        })

                return {
                    "success": True,
                    "target_date": target_date_str,
                    "total_operations": len(operations),
                    "success_count": success_count,
                    "failed_count": failed_count,
                    "results": results
                }

        except Exception as e:
            print(f"Batch evaluation error: {e}", file=sys.stderr, flush=True)
            return {
                "success": False,
                "error": str(e)
            }


def main():
    """Main entry point for CLI usage"""
    input_data = read_input()

    evaluator = OperationEvaluator()

    action = input_data.get('action')

    if action == 'evaluate':
        # Evaluate single operation
        result = evaluator.evaluate_operation(
            operation_id=input_data.get('operationId')
        )
    elif action == 'leaderboard':
        # Get optimizer leaderboard
        result = evaluator.get_optimizer_leaderboard(
            days=input_data.get('days', 30),
            limit=input_data.get('limit', 20)
        )
    elif action == 'evaluate_7days_ago':
        # Batch evaluate operations from 7 days ago
        result = evaluator.evaluate_operations_7days_ago()
    else:
        result = {"error": f"Unknown action: {action}"}

    format_output(result)


if __name__ == "__main__":
    main()
